doc/plutus/architecture/constraints/index.rst

```diff
@@ -0,0 +1,294 @@
+More specifically, it describes
+
+* how the constraint-based API is used to generate Plutus scripts and transactions.
+* the rational behind the selected approach
+* possible alternative approaches
+
+Specification
+-------------
+
+We define the constraints with the following data types.
+
+.. code-block:: haskell
+
+  data TxConstraints i o =
+      TxConstraints
+          { txConstraints :: [TxConstraint]
+          , txOwnInputs   :: [ScriptInputConstraint i]
+          , txOwnOutputs  :: [ScriptOutputConstraint o]
+          }
+
+  data TxConstraint =
+        MustHashDatum DatumHash Datum
+      -- ^ The transaction's datum witnesses must contain the given 'DatumHash'
+      -- and 'Datum'. Useful when you already have a 'DatumHash' and
+      -- want to make sure that it is the actual hash of the 'Datum'.
+      | MustIncludeDatum Datum
+      -- ^ Like 'MustHashDatum', but the hash of the 'Datum' is computed automatically.
+      | MustValidateIn POSIXTimeRange
+      -- ^ The transaction's validity range must be set with the given 'POSIXTimeRange'.
+      | MustBeSignedBy PaymentPubKeyHash
+      -- ^ The transaction must add the given 'PaymentPubKeyHash' in its signatories.
+      | MustSpendAtLeast Value
+      -- ^ The sum of the transaction's input 'Value's must be at least as much as
+      -- the given 'Value'.
+      | MustProduceAtLeast Value
+      -- ^ The sum of the transaction's output 'Value's must be at least as much as
+      -- the given 'Value'.
+      | MustSpendPubKeyOutput TxOutRef
+      -- ^ The transaction must spend the given unspent transaction public key output.
+      | MustSpendScriptOutput TxOutRef Redeemer
+      -- ^ The transaction must spend the given unspent transaction script output.
+      | MustMintValue MintingPolicyHash Redeemer TokenName Integer
+      -- ^ The transaction must mint the given token and amount.
+      | MustPayToPubKeyAddress PaymentPubKeyHash (Maybe StakePubKeyHash) (Maybe Datum) Value
+      -- ^ The transaction must create a transaction output with a public key address.
+      | MustPayToOtherScript ValidatorHash Datum Value
+      -- ^ The transaction must create a transaction output with a script address.
+      | MustSatisfyAnyOf [[TxConstraint]]
+
+  data ScriptInputConstraint a =
+      ScriptInputConstraint
+          { icRedeemer :: a -- ^ The typed 'Redeemer' to be used with the target script
+          , icTxOutRef :: TxOutRef -- ^ The UTXO to be spent by the target script
+          }
+
+  data ScriptOutputConstraint a =
+      ScriptOutputConstraint
+          { ocDatum :: a -- ^ Typed datum to be used with the target script
+          , ocValue :: Value
+          }
+
+The `TxConstraints` can be combined with its `Semigroup` instance.
+Values of `TxConstraints` are constructed using smart constructors like `mustBeSignedBy`, `mustPayToTheScript`, `mustSpendAtLeast`, etc. which are available in Ledger.Constraints.TxConstraints_.
```

Comment:
1. perhaps use "obtained" rather than "constructed" to avoid the repetition with "constructor".
2. Is the `TxConstraints` constructor exported at all?
3. Perhaps "Values of `TxConstraints` need to be obtained using smart constructors like ..." ?

i.e. "are constructed usng smart constructors" or "have to be